### PR TITLE
Edpub 1397 & EDPUB-1398

### PR DESCRIPTION
**Pull request by acharya-deepak**
# Description
1. Create & Edit API Endpoint for adding an input
2. Create an API Endpoint for editing a question
## Spec

Designs: 
See Ticket:[EDPUB-1397](https://bugs.earthdata.nasa.gov/browse/EDPUB-1397)
[EDPUB-1398](https://bugs.earthdata.nasa.gov/browse/EDPUB-1398)
[EDPUB-1396](https://bugs.earthdata.nasa.gov/browse/EDPUB-1396)

---

## Validation
**for EDPBUB-1397 and EDPUB-1398**
1. In swagger, to test create: look for api/data/input/create 
sample data: make sure question_id exists as its a foreign key

{
  "question_id": "8a364184-42ac-48fe-b831-acb2eb08c729",
  "control_id": "data_producers_table",
  "list_order": 0,
  "label": "",
  "type": "table",
  "enums": "[{\"key\": \"producer_first_name\", \"type\": \"text\", \"label\": \"First Name\", \"editable\": true}, {\"key\": \"producer_middle_initial\", \"type\": \"text\", \"label\": \"Middle Initial\", \"editable\": true}, {\"key\": \"producer_last_name_or_organization\", \"type\": \"text\", \"label\": \"Last Name or Group\", \"editable\": true}]",
  "attributes": "{}",
  "required_if": "[]",
  "show_if": "[]",
  "required": true
}

2. In swagger, to test update: look for api/data/input/update

{
  "question_id": "8a364184-42ac-48fe-b831-acb2eb08c729",
  "control_id": "data_producers_table",
  "list_order": 1,
  "label": "",
  "type": "table",
  "enums": "[{\"key\": \"producer_first_name\", \"type\": \"text\", \"label\": \"First Name\", \"editable\": true}, {\"key\": \"producer_middle_initial\", \"type\": \"text\", \"label\": \"Middle Initial\", \"editable\": true}, {\"key\": \"producer_last_name_or_organization\", \"type\": \"text\", \"label\": \"Last Name or Group\", \"editable\": true}]",
  "attributes": "{}",
  "required_if": "[]",
  "show_if": "[]",
  "required": true
}

here list_order is updated from 0 to1

3. Added data/inputs API to retrieve all the input rows.

4. Added data/input/{id}/{controlId} API to retrieve the row with question_id and control_id

**For EDPUB-1393:**
use API /data/question/{id}

{
  "id": "f3e2eab9-6375-4e53-9cc2-3d16f318d333",
  "short_name": "point_of_contact",
  "version": 1,
  "long_name": "Data Accession Point of Contact**222**",
  "text": "Who should the DAAC contact with questions regarding this Data Accession request?",
  "help": "This person should have in-depth knowledge of this data product, allowing them to provide additional information as needed.",
  "required": false,
  "created_at": "2024-11-04T19:28:02.054Z",
  "daac_ids": ["40397fe8-4841-4e4c-b84a-6ece359ff5ff"],
  "inputs": [
    {
      "type": "text",
      "enums": {},
      "label": "Full Name",
      "show_if": [],
      "required": true,
      "attributes": {},
      "control_id": "poc_name",
      "list_order": 0,
      "required_if": []
    },
    {
      "type": "text",
      "enums": {},
      "label": "Organization",
      "show_if": [],
      "required": true,
      "attributes": {},
      "control_id": "poc_organization",
      "list_order": 1,
      "required_if": []
    },
    {
      "type": "text",
      "enums": {},
      "label": "Department",
      "show_if": [],
      "required": false,
      "attributes": {},
      "control_id": "poc_department",
      "list_order": 2,
      "required_if": []
    },
    {
      "type": "email",
      "enums": {},
      "label": "Email",
      "show_if": [],
      "required": true,
      "attributes": {},
      "control_id": "poc_email",
      "list_order": 3,
      "required_if": []
    },
    {
      "type": "text",
      "enums": {},
      "label": "ORCID",
      "show_if": [],
      "required": false,
      "attributes": {},
      "control_id": "poc_orcid",
      "list_order": 4,
      "required_if": []
    }
  ]
}

### CODE REVIEW

File: src/nodejs/controllers/proxy.js [hunk 1]
```diff
@@ -84,8 +84,16 @@ module.exports.formFindById = function formFindById(req, res, next) {
 };
 
 module.exports.formPut = function formPut(req, res, next) {
```
Comment (camposeddie) [inline src/nodejs/controllers/proxy.js:86]: I think this is only included in the PR due to creating this branch off the EDPUB-1391 branch. In the future, it'll be best just to branch off develop then work through merge conflicts as they pop up.
Hopefully that will help isolate changes and make review a bit easier

Comment (acharya-deepak) [reply src/nodejs/controllers/proxy.js:86]: Yes, I agree, and I will keep this in mind next time.

File: src/nodejs/lambda-layers/database-util/src/query/question.js [hunk 2]
```diff
@@ -114,16 +114,22 @@ const findAllEx = () => `
...
-  VALUES ({{payload.id}}, {{payload.short_name}}, {{payload.version}}, {{payload.long_name}}, {{payload.text}},
-  {{payload.help}}, {{payload.required}}, {{payload.created_at}}, ARRAY(
-    SELECT DISTINCT unnest(array_cat(daac_ids, ARRAY['${daac_ids.join('\',\'')}']::UUID[]))
-  ))
-  ON CONFLICT(short_name, version) DO UPDATE SET
-  long_name = EXCLUDED.long_name, text = EXCLUDED.text, help = EXCLUDED.help,
-  required = EXCLUDED.required, created_at = EXCLUDED.created_at, EXCLUDED.daac_ids
-  RETURNING *`;
+const update = (params) => 
+  `INSERT INTO question (id, short_name, version, long_name, text, help, required, created_at, daac_ids)
```
Comment (ckd002) [inline src/nodejs/lambda-layers/database-util/src/query/question.js:118]: This throws a duplicate key error if you attempt to update the short_name or version number.

Comment (acharya-deepak) [reply src/nodejs/lambda-layers/database-util/src/query/question.js:118]: I have updated the query and checked for only the unique key in the new query.

File: src/openapi/openapi.json [hunk 3]
```diff
@@ -2522,6 +2522,512 @@
...
+            "description": "Bad request"
+          }
+        },
+        "security": [
+          {
+            "Cognito_Authorizer": ["openid"]
+          }
+        ],
+        "x-amazon-apigateway-integration": {
+          "uri": "${data_lambda_arn}",
```
Comment (camposeddie) [inline src/openapi/openapi.json:2571]: This is pointing to the data lambda where it should be pointed to the question lambda

File: src/openapi/openapi.json [hunk 4]
```diff
@@ -2522,6 +2522,512 @@
...
+              "responseTemplates": {
+                "application/json": "{}"
+              }
+            }
+          }
+        },
+        "x-router-controller": "cors"
+      }
+    },
+    "/api/data/input/create": {
```
Comment (camposeddie) [inline src/openapi/openapi.json:2781]: This should be a POST to the `/api/data/inputs` endpoint.

File: src/openapi/openapi.json [hunk 5]
```diff
@@ -2522,6 +2522,512 @@
...
+              "responseTemplates": {
+                "application/json": "{}"
+              }
+            }
+          }
+        },
+        "x-router-controller": "cors"
+      }
+    },  
+    "/api/data/input/update": {
```
Comment (camposeddie) [inline src/openapi/openapi.json:2906]: This should be a POST to the `/api/data/input/{id}/{controlId}` endpoint

**Review by camposeddie — CHANGES_REQUESTED** [review]
API naming changes requested

**Review by camposeddie — APPROVED** [review]

**Pull request merged**